### PR TITLE
[memprof] Don't use Frame::hash or hashCallStacks in unit test

### PR DESCRIPTION
**Pull request by kazutakahirata**
This patch checks the result of YAML parsing at the level of
MemProfRecord instead of IndexedMemProfRecord, thereby avoiding use of
Frame::hash and hashCallStacks.  This makes sense because we
ultimately care about consumers like MemProfiler.cpp obtaining
MemProfRecord correctly; IndexedMemProfData and hash values are just
intermediaries.

Once this patch lands, we call Frame::hash and hashCallStacks only
when adding Frames or call stacks to their respective data structures.
In other words, the hash functions are pretty much business internal
to IndexedMemProfRecord.

### CODE REVIEW

**Comment by kazutakahirata** [comment]
@teresajohnson @snehasish Friendly ping.  Thanks!

File: llvm/unittests/ProfileData/MemProfTest.cpp [hunk 1]
```diff
@@ -709,47 +709,33 @@ TEST(MemProf, YAMLParser) {
...
-
   // Verify the entire contents of MemProfData.Records.
   ASSERT_THAT(MemProfData.Records, SizeIs(1));
-  const auto &[GUID, Record] = MemProfData.Records.front();
+  const auto &[GUID, IndexedRecord] = MemProfData.Records.front();
   EXPECT_EQ(GUID, 0xdeadbeef12345678ULL);
+
+  FrameIdConverter<decltype(MemProfData.Frames)> FrameIdConv(
+      MemProfData.Frames);
+  CallStackIdConverter<decltype(MemProfData.CallStacks)> CSIdConv(
```
Comment (teresajohnson) [inline llvm/unittests/ProfileData/MemProfTest.cpp:719]: This code sequence seems to exist in a lot of places within this test and maybe others - consider refactoring into a helper in a follow on PR?

**Review by teresajohnson — APPROVED** [review]

File: llvm/unittests/ProfileData/MemProfTest.cpp (hunk 1 above)
Comment (kazutakahirata) [reply llvm/unittests/ProfileData/MemProfTest.cpp:719]: Thank you for pointing this out!  Will do.
